File: xwcd/ds/lin/id.c
/**
 * @file
 * @brief 玄武设备栈：LIN总线ID校验表
 * @author
 * + 隐星魂 (Roy Sun) <xwos@xwos.tech>
 * @copyright
 * + Copyright © 2015 xwos.tech, All Rights Reserved.
 * > Licensed under the Apache License, Version 2.0 (the "License");
 * > you may not use this file except in compliance with the License.
 * > You may obtain a copy of the License at
 * >
 * >         http://www.apache.org/licenses/LICENSE-2.0
 * >
 * > Unless required by applicable law or agreed to in writing, software
 * > distributed under the License is distributed on an "AS IS" BASIS,
 * > WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * > See the License for the specific language governing permissions and
 * > limitations under the License.
 */

#include <xwcd/ds/standard.h>
#include <xwcd/ds/lin/id.h>

/**
 * @brief LIN总线ID校验表
 */
__xwds_rodata const xwu8_t xwds_linid_table[64] = {
        LIN_ID_PARITY_FIELD(0), /* 0 */
        LIN_ID_PARITY_FIELD(1), /* 1 */
        LIN_ID_PARITY_FIELD(2), /* 2 */
        LIN_ID_PARITY_FIELD(3), /* 3 */
        LIN_ID_PARITY_FIELD(4), /* 4 */
        LIN_ID_PARITY_FIELD(5), /* 5 */
        LIN_ID_PARITY_FIELD(6), /* 6 */
        LIN_ID_PARITY_FIELD(7), /* 7 */
        LIN_ID_PARITY_FIELD(8), /* 8 */
        LIN_ID_PARITY_FIELD(9), /* 9 */
        LIN_ID_PARITY_FIELD(10), /* 10 */
        LIN_ID_PARITY_FIELD(11), /* 11 */
        LIN_ID_PARITY_FIELD(12), /* 12 */
        LIN_ID_PARITY_FIELD(13), /* 13 */
        LIN_ID_PARITY_FIELD(14), /* 14 */
        LIN_ID_PARITY_FIELD(15), /* 15 */
        LIN_ID_PARITY_FIELD(16), /* 16 */
        LIN_ID_PARITY_FIELD(17), /* 17 */
        LIN_ID_PARITY_FIELD(18), /* 18 */
        LIN_ID_PARITY_FIELD(19), /* 19 */
        LIN_ID_PARITY_FIELD(20), /* 20 */
        LIN_ID_PARITY_FIELD(21), /* 21 */
        LIN_ID_PARITY_FIELD(22), /* 22 */
        LIN_ID_PARITY_FIELD(23), /* 23 */
        LIN_ID_PARITY_FIELD(24), /* 24 */
        LIN_ID_PARITY_FIELD(25), /* 25 */
        LIN_ID_PARITY_FIELD(26), /* 26 */
        LIN_ID_PARITY_FIELD(27), /* 27 */
        LIN_ID_PARITY_FIELD(28), /* 28 */
        LIN_ID_PARITY_FIELD(29), /* 29 */
        LIN_ID_PARITY_FIELD(30), /* 30 */
        LIN_ID_PARITY_FIELD(31), /* 31 */
        LIN_ID_PARITY_FIELD(32), /* 32 */
        LIN_ID_PARITY_FIELD(33), /* 33 */
        LIN_ID_PARITY_FIELD(34), /* 34 */
        LIN_ID_PARITY_FIELD(35), /* 35 */
        LIN_ID_PARITY_FIELD(36), /* 36 */
        LIN_ID_PARITY_FIELD(37), /* 37 */
        LIN_ID_PARITY_FIELD(38), /* 38 */
        LIN_ID_PARITY_FIELD(39), /* 39 */
        LIN_ID_PARITY_FIELD(40), /* 40 */
        LIN_ID_PARITY_FIELD(41), /* 41 */
        LIN_ID_PARITY_FIELD(42), /* 42 */
        LIN_ID_PARITY_FIELD(43), /* 43 */
        LIN_ID_PARITY_FIELD(44), /* 44 */
        LIN_ID_PARITY_FIELD(45), /* 45 */
        LIN_ID_PARITY_FIELD(46), /* 46 */
        LIN_ID_PARITY_FIELD(47), /* 47 */
        LIN_ID_PARITY_FIELD(48), /* 48 */
        LIN_ID_PARITY_FIELD(49), /* 49 */
        LIN_ID_PARITY_FIELD(50), /* 50 */
        LIN_ID_PARITY_FIELD(51), /* 51 */
        LIN_ID_PARITY_FIELD(52), /* 52 */
        LIN_ID_PARITY_FIELD(53), /* 53 */
        LIN_ID_PARITY_FIELD(54), /* 54 */
        LIN_ID_PARITY_FIELD(55), /* 55 */
        LIN_ID_PARITY_FIELD(56), /* 56 */
        LIN_ID_PARITY_FIELD(57), /* 57 */
        LIN_ID_PARITY_FIELD(58), /* 58 */
        LIN_ID_PARITY_FIELD(59), /* 59 */
        LIN_ID_PARITY_FIELD(60), /* 60 */
        LIN_ID_PARITY_FIELD(61), /* 61 */
        LIN_ID_PARITY_FIELD(62), /* 62 */
        LIN_ID_PARITY_FIELD(63), /* 63 */
};

File: xwam/application/w25qpt/py/id.py
#! /usr/bin/python3
# -*- coding:utf-8 -*-
# @file
# @brief W25Qxx编程器：ID读取工具
# @author
# + 隐星曜 (Roy Sun) <xwos@xwos.tech>
# @copyright
# + Copyright © 2015 xwos.tech, All Rights Reserved.
# > Licensed under the Apache License, Version 2.0 (the "License");
# > you may not use this file except in compliance with the License.
# > You may obtain a copy of the License at
# >
# >         http://www.apache.org/licenses/LICENSE-2.0
# >
# > Unless required by applicable law or agreed to in writing, software
# > distributed under the License is distributed on an "AS IS" BASIS,
# > WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# > See the License for the specific language governing permissions and
# > limitations under the License.
#

import argparse
from hwif import hwif

msg_read_uid = bytearray([0xAA, 0xAA, 0xAA, 0xAA,
                          0x00, 0x00, 0x00, 0x00,
                          0x00, 0x00, 0x00, 0x00,
                          0x65, 0x22, 0xDF, 0x69,
                          0x55, 0x55])

msg_read_mid = bytearray([0xAA, 0xAA, 0xAA, 0xAA,
                          0x01, 0x00, 0x00, 0x00,
                          0x00, 0x00, 0x00, 0x00,
                          0xA9, 0x88, 0xDF, 0xF7,
                          0x55, 0x55])

msg_read_jid = bytearray([0xAA, 0xAA, 0xAA, 0xAA,
                          0x02, 0x00, 0x00, 0x00,
                          0x00, 0x00, 0x00, 0x00,
                          0x27, 0x07, 0xD8, 0x14,
                          0x55, 0x55])


def read_uid(port: str, baudrate: int):
    with hwif.HwIf(port = port, baudrate = baudrate) as hwport:
        txsz = hwport.tx(msg_read_uid)
        if txsz == len(msg_read_uid):
            rxb = hwport.rx()
            if not rxb == None:
                oprc = rxb[5]
                if (oprc == 0):
                    print(rxb[12:20].hex().upper())
                elif (oprc == 1):
                    print("读取UID... -EINTR")
                elif (oprc == 2):
                    print("读取UID... -ETIMEDOUT")
                else:
                    print("读取UID... -EIO")


def read_mid(port: str, baudrate: int):
    with hwif.HwIf(port = port, baudrate = baudrate) as hwport:
        txsz = hwport.tx(msg_read_mid)
        if txsz == len(msg_read_mid):
            rxb = hwport.rx()
            if not rxb == None:
                oprc = rxb[5]
                if (oprc == 0):
                    print(rxb[12:14].hex().upper())
                elif (oprc == 1):
                    print("读取MID... -EINTR")
                elif (oprc == 2):
                    print("读取MID... -ETIMEDOUT")
                else:
                    print("读取MID... -EIO")


def read_jid(port: str, baudrate: int):
    with hwif.HwIf(port = port, baudrate = baudrate) as hwport:
        txsz = hwport.tx(msg_read_jid)
        if txsz == len(msg_read_jid):
            rxb = hwport.rx()
            if not rxb == None:
                oprc = rxb[5]
                if (oprc == 0):
                    print(rxb[13:16].hex().upper())
                elif (oprc == 1):
                    print("读取JID... -EINTR")
                elif (oprc == 2):
                    print("读取JID... -ETIMEDOUT")
                else:
                    print("读取JID... -EIO")


def parse_arg():
    parser = argparse.ArgumentParser(description="W25Qxx编程器：ID读取工具")
    parser.add_argument("-p", "--port",
                        action="store", dest="port", default="/dev/ttyUSB0",
                        help="串口端口")
    parser.add_argument("-b", "--baudrate",
                        action="store", dest="baudrate", type=int, default=1000000,
                        help="波特率")
    parser.add_argument('id',
                        metavar='ID',
                        help="ID类型：uid,mid,jid")
    return parser


def main():
    parser = parse_arg()
    args = parser.parse_args()
    if args.id == "uid":
        read_uid(args.port, args.baudrate)
    elif args.id == "mid":
        read_mid(args.port, args.baudrate)
    elif args.id == "jid":
        read_jid(args.port, args.baudrate)

if __name__ == "__main__":
    main()
